### PR TITLE
[#3141]Move the related notify test cases code to common module, fix NPE and improve peformance.

### PR DESCRIPTION
**Pull request by zongtanghu**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

(1)Move the related notify test cases code to common module.
(2)fix the issue that smartsubscriber cause NPE.

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [x] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [x] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

**Comment by zongtanghu** [comment]
In this pr, there are two points:
(1)fix the issue that smartsubscriber cause NPE(I unified smartsubscriber and subssriber, it cause a NPE Because smartsubscriber's subscribeType return null by default).
(2)Move the related notify test cases code to common module.(But original test cases codes is still reservered in the core module project.)
@yanlinly @chuntaojun @KomachiSion

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java [hunk 1]
```diff
@@ -175,10 +175,16 @@ void receiveEvent(Event event) {
...
             
-            final String targetName = ClassUtils.getName(subscriber.subscribeType());
-            if (!Objects.equals(sourceName, targetName)) {
-                continue;
+            // Because unifying smartSubscriber and subscriber, so here need to think of compatibility.
+            // To smartSubscriber, the subscribeType() method returns null by default.
+            if (subscriber.subscribeType() != null) {
+                
+                final String targetName = ClassUtils.getName(subscriber.subscribeType());
+                if (!Objects.equals(sourceName, targetName)) {
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java:183]: 这个的存在是因为sharePublisher，建议sharePublisher重写DefaultPublisher的receiveEvent方法，比对事件的签名信息，避免出现ClassCastException

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java [hunk 2]
```diff
@@ -42,21 +40,21 @@
  */
 public class DefaultPublisher extends Thread implements EventPublisher {
     
-    private static final Logger LOGGER = LoggerFactory.getLogger(NotifyCenter.class);
+    protected static final Logger LOGGER = LoggerFactory.getLogger(NotifyCenter.class);
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java:43]: Why change it to protected? It seems that there are no subClass.

Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java:43]: I implement DefaultSharePublisher which is a subclass of DefaultPublisher  to separate sharePublish from defaultPublisher.

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import com.alibaba.nacos.common.notify.listener.Subscriber;
+import com.alibaba.nacos.common.utils.ClassUtils;
+import com.alibaba.nacos.common.utils.Objects;
+
+/**
+ * The default share event publisher implementation for slow event.
+ *
+ * @author zongtanghu
+ */
+public class DefaultSharePublisher extends DefaultPublisher {
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java:29]: 或许这里可以在加速下性能，Map<Class<? extends SlowEvent>, Set<Subscriber>>, 这样就不需要每个都去equals

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        // Notification single event listener
+        for (Subscriber subscriber : subscribers) {
+            // Whether to ignore expiration events
+            if (subscriber.ignoreExpireEvent() && lastEventSequence > currentEventSequence) {
+                LOGGER.debug("[NotifyCenter] the {} is unacceptable to this subscriber, because had expire",
+                        event.getClass());
+                continue;
+            }
+            
+            final String targetName = ClassUtils.getName(subscriber.subscribeType());
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java]: The code has already adjusted by previous commit.

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java (hunk 3 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java:31]: The part codes of improving peformance has already been committed.

**Comment by zongtanghu** [comment]
The CI has already been passed, please help to review, thanks. @chuntaojun

**Comment by zongtanghu** [comment]
The CI has already been passed, please help to review, thanks. @chuntaojun

File: common/src/test/java/com/alibaba/nacos/common/notify/NotifyCenterTest.java [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+            return System.currentTimeMillis();
+        }
+    }
+    
+    static {
+        System.setProperty("nacos.core.notify.share-buffer-size", "8");
+    }
+    
+    @Test
+    public void test_a_event_can_listen() throws Exception {
```
Comment (KomachiSion) [inline common/src/test/java/com/alibaba/nacos/common/notify/NotifyCenterTest.java:57]: ut test method also use Camel Naming

**Comment by zongtanghu** [comment]
Okay, I will adjust method torromow and push a commit again!


 原始邮件 
发件人: 杨翊 SionYang<notifications@github.com>
收件人: alibaba/nacos<nacos@noreply.github.com>
抄送: Hu Zongtang<huzongtang@cmss.chinamobile.com>; Author<author@noreply.github.com>
发送时间: 2020年6月23日(周二) 22:18
主题: Re: [alibaba/nacos] [#3141]Move the related notify test cases code tocommon module, fix NPE and improve peformance. (#3145)


@KomachiSion requested changes on this pull request.
In common/src/test/java/com/alibaba/nacos/common/notify/NotifyCenterTest.java:
> + } + + private static class TestEvent extends Event { + + @Override + public long sequence() { + return System.currentTimeMillis(); + } + } + + static { + System.setProperty("nacos.core.notify.share-buffer-size", "8"); + } + + @Test + public void test_a_event_can_listen() throws Exception {
ut test method also use Camel Naming

File: common/src/test/java/com/alibaba/nacos/common/notify/NotifyCenterTest.java (hunk 5 above)
Comment (zongtanghu) [reply common/src/test/java/com/alibaba/nacos/common/notify/NotifyCenterTest.java]: Yeah, I have already renamed these ut test method name in next commit, please review again, thanks.

**Comment by zongtanghu** [comment]
@chuntaojun @KomachiSion Please help to review again, thanks very much.

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java [hunk 6]
```diff
@@ -125,12 +127,12 @@ private boolean hasSubscriber() {
     }
     
     @Override
-    public void addSubscriber(Subscriber subscriber) {
+    public void addSubscriber(Subscriber subscriber, Class<? extends Event> subscribeType) {
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java:130]: 这应该是DefaultSharePublisher特有的方法才是

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java [hunk 7]
```diff
@@ -162,9 +164,13 @@ public boolean isInitialized() {
         return initialized;
     }
     
-    void receiveEvent(Event event) {
+    /**
+     * Receive and notifySubscriber to process the event.
+     *
+     * @param event {@link Event}.
+     */
+    public void receiveEvent(Event event) {
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java:172]: 这里为什么要调整为public修饰？

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        // Actually, do a classification based on the slowEvent type.
+        Class<? extends SlowEvent> subSlowEventType = (Class<? extends SlowEvent>) subscribeType;
+        // For removing to parent class attributes synchronization.
+        subscribers.remove(subscriber);
+    
+        Set<Subscriber> sets = subMappings.get(subSlowEventType);
+    
+        if (sets != null && sets.contains(subscriber)) {
+            sets.remove(subscriber);
+            if (sets.isEmpty()) {
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java:64]: 这里并非原子操作，需要注意

File: common/src/main/java/com/alibaba/nacos/common/notify/EventPublisher.java [hunk 9]
```diff
@@ -46,15 +46,17 @@
      * Add listener.
      *
      * @param subscriber {@link Subscriber}
+     * @param subscribeType type of subscribe
      */
-    void addSubscriber(Subscriber subscriber);
+    void addSubscriber(Subscriber subscriber, Class<? extends Event> subscribeType);
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/notify/EventPublisher.java:51]: 同上，这个接口本不是一个公共接口，他只是适用于SharePublisher

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java (hunk 6 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java:130]: Okay, I will adjust this codes.

File: common/src/main/java/com/alibaba/nacos/common/notify/EventPublisher.java (hunk 9 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/EventPublisher.java:51]: Okay, I will adjust this codes.

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java (hunk 7 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java:172]: I will revert this part.

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+     *
+     * @param subscriber {@link Subscriber}
+     * @param subscribeType subscribe event type, such as slow event or general event.
+     */
+    public void addSubscriber(Subscriber subscriber, Class<? extends Event> subscribeType) {
+        // Actually, do a classification based on the slowEvent type.
+        Class<? extends SlowEvent> subSlowEventType = (Class<? extends SlowEvent>) subscribeType;
+        // For adding to parent class attributes synchronization.
+        subscribers.add(subscriber);
+        Set<Subscriber> sets = subMappings.get(subSlowEventType);
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java:46]: 这里可能还是会有原子性的问题，可能存在两个线程同时去addSubscriber，get后的结果同时为null

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+     * Add listener for default share publisher.
+     *
+     * @param subscriber {@link Subscriber}
+     * @param subscribeType subscribe event type, such as slow event or general event.
+     */
+    public void addSubscriber(Subscriber subscriber, Class<? extends Event> subscribeType) {
+        // Actually, do a classification based on the slowEvent type.
+        Class<? extends SlowEvent> subSlowEventType = (Class<? extends SlowEvent>) subscribeType;
+        // For adding to parent class attributes synchronization.
+        subscribers.add(subscriber);
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java:45]: 这里应该不再需要调用 subscribers.add(subscriber)

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+     * Remove listener for default share publisher.
+     *
+     * @param subscriber {@link Subscriber}
+     * @param subscribeType subscribe event type, such as slow event or general event.
+     */
+    public void removeSubscriber(Subscriber subscriber, Class<? extends Event> subscribeType) {
+        // Actually, do a classification based on the slowEvent type.
+        Class<? extends SlowEvent> subSlowEventType = (Class<? extends SlowEvent>) subscribeType;
+        // For removing to parent class attributes synchronization.
+        subscribers.remove(subscriber);
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java:68]: 同理，这个也不再需要调用了

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java (hunk 11 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java:45]: I suggest to retain.

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java (hunk 12 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java:68]: I suggest to retain.

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java (hunk 8 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java]: I revert this codes and fix the other same issue.

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java (hunk 10 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/DefaultSharePublisher.java:46]: I will fix and resolve the same other issue.

**Comment by zongtanghu** [comment]
@KomachiSion @yanlinly  please help to review this pr.

**Pull request merged**